### PR TITLE
feat: add documentation for go sdk

### PR DESCRIPTION
**Pull request by danijelTxFusion**

### CODE REVIEW

File: docs/api/go/README.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Go SDK
+
+While most of the existing SDKs should work out of the box, deploying smart contracts or using unique zkSync features, like paying fees in other tokens, requires providing additional fields to those that Ethereum transactions have by default.
```
Comment (kailash-manasarovar) [inline docs/api/go/README.md:3]:
```suggestion
While most of the existing SDKs should work out of the box, deploying smart contracts or using unique zkSync features, such as paying fees in other tokens, requires providing additional fields to those that Ethereum transactions have by default.
```

File: docs/api/go/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+To provide easy access to all the features of zkSync Era, the `zksync2` Python SDK was created.
```
Comment (kailash-manasarovar) [inline docs/api/go/README.md:5]:
```suggestion
To provide easy access to all the features of zkSync Era, the `zksync2` Go SDK was created.
```

File: docs/api/go/README.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+## Table of contents
+
+- [Getting started](./getting-started.md)
+- [Providers](./providers.md)
+- [Accounts: Overview](./accounts.md)
+- [Accounts: L1->L2 transactions](./accounts-l1-l2.md)
+- [Contracts Interface](./contract-interface.md)
```
Comment (kailash-manasarovar) [inline docs/api/go/README.md:13]:
```suggestion
- [Contract interface](./contract-interface.md)
```

File: docs/api/go/accounts-l1-l2.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
+# Accounts: L1->L2 transactions
+
+This section explores the methods which allow the [account](./accounts.md) classes to send transactions from L1 to L2.
+
+If you want to get some background on how L1->L2 interaction works on zkSync, go through the [introduction](../../dev/developer-guides/bridging/l1-l2-interop.md).
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:5]:
```suggestion
If you want some background on how L1->L2 interaction works on zkSync, go through the [introduction](../../dev/developer-guides/bridging/l1-l2-interop.md).
```

File: docs/api/go/accounts-l1-l2.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+Full examples of actions below are available at [Getting Started](./getting-started.md) page.
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:7]:
```suggestion
Full examples of actions below are available on the [getting started](./getting-started.md) page.
```

File: docs/api/go/accounts-l1-l2.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+## Deposit
+
+```go
+
+func (p *DefaultEthProvider) Deposit(token *Token, amount *big.Int, address common.Address, options *GasOptions) (*types.Transaction, error)
+
+```
+
+**Inputs and Outputs**
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:17]:
```suggestion
**Inputs and outputs**
```

File: docs/api/go/accounts-l1-l2.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+| Name    | Description                                               |
+|---------|-----------------------------------------------------------|
+| token   | The address of the deposited L1 ERC20 token.              |
+| amount  | The amount of the token to be deposited.                  |
+| address | The address that will receive the deposited tokens on L2. |
+| options | The options for transaction fee and gas.                  |
+| returns | The transaction hash of the deposit.                      |
+
+
+## Claim Failed Deposit
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:28]:
```suggestion
## Claim failed deposit
```

File: docs/api/go/accounts-l1-l2.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+The `ClaimFailedDeposit` method withdraws funds from the initiated deposit, which failed when finalizing on L2.  
+If the deposit L2 transaction has failed, it sends an L1 transaction calling `ClaimFailedDeposit` method of the
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:31]:
```suggestion
If the L2 deposit transaction fails, it sends an L1 transaction calling the `ClaimFailedDeposit` method of the 
```

File: docs/api/go/accounts-l1-l2.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+Withdrawal are executed in 2 steps: 
+ - `Withdraw`: Initiates withdrawal on L2. 
+ - `FinalizeWithdraw`: Finalized withdrawal on L1.
+
+### Withdraw
+
+```go
+func (w *Wallet) Withdraw(to common.Address, amount *big.Int, token *Token, nonce *big.Int) (common.Hash, error)
+```
+**Inputs and Outputs**
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:60]:
```suggestion
**Inputs and outputs**
```

File: docs/api/go/accounts-l1-l2.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+| Name    | Description                                              |
+|---------|----------------------------------------------------------|
+| to      | The address that will receive the withdrew tokens on L1. |
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:64]:
```suggestion
| to      | The address that receives the withdraw tokens on L1. |
```

File: docs/api/go/accounts-l1-l2.md [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+| amount  | The amount of the token to be deposited.                 |
+| token   | The address of the L2 ERC20 token.                       |
+| nonce   | The sender's nonce.                                      |
+| returns | The transaction hash of the withdrawal.                  |
+
+
+
+### FinalizeWithdraw
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:72]:
```suggestion
### `FinalizeWithdraw`
```

File: docs/api/go/accounts-l1-l2.md [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+```go
+func (w *Wallet) FinalizeWithdraw(withdrawalHash common.Hash, index int) (common.Hash, error)
+```
+
+**Inputs and Outputs**
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:78]:
```suggestion
**Inputs and outputs**
```

File: docs/api/go/accounts-l1-l2.md [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+| Name           | Description                                                                          |
+|----------------|--------------------------------------------------------------------------------------|
+| withdrawalHash | The L2 transaction hash of the withdrawal.                                           |
+| index          | The position in the L2 logs Merkle tree of the l2Log that was sent with the message. |
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:83]:
```suggestion
| index          | The position in the L2 Merkle-tree logs of the `l2Log` that was sent with the message. |
```

File: docs/api/go/accounts-l1-l2.md [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+| returns        | The L1 transaction hash of the finilized withdrawal.                                 |
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts-l1-l2.md:84]:
```suggestion
| returns        | The L1 transaction hash of the finalized withdrawal.                                 |
```

File: docs/api/go/accounts.md [hunk 15]
```diff
@@ -0,0 +1,95 @@
+# Accounts: overview
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts.md:1]:
```suggestion
# Accounts: Overview
```

File: docs/api/go/accounts.md [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+`DefaultEthSigner` provides private key management, transaction and message
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts.md:3]:
```suggestion
`DefaultEthSigner` provides private key management, transaction, and message 
```

File: docs/api/go/accounts.md [hunk 17]
```diff
@@ -0,0 +1,95 @@
+# Accounts: overview
+
+`DefaultEthSigner` provides private key management, transaction and message 
+signing.
+
+## DefaultEthSigner
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts.md:6]:
```suggestion
## `DefaultEthSigner`
```

File: docs/api/go/accounts.md [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+- `SignHash`: signs transaction hash.
+- `SignTypedData`: signs EIP712-typed zkSync transaction which is used 
+for smart contract deployment.
+ 
+
+# Wallet
+
+A `Wallet` is a wrapper around `EthSigner` which provides performing actions on L2 network.
+Based on the action, the `Wallet` crates an appropriate transaction, signs the transaction using
+the `EthSigner` and then broadcasts the transaction to the network.
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts.md:59]:
```suggestion
the `EthSigner`, and then broadcasts the transaction to the network.
```

File: docs/api/go/accounts.md [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+// Create wallet
+w, err := zksync2.NewWallet(es, zp)
+if err != nil {
+	log.Panic(err)
+}
+```
+Visit [Accounts:L1->L2 Transactions](./accounts-l1-l2.md) page to see which actions `Wallet` provides.
+
+Complete examples are available on [Getting Started](./getting-started.md) page.
```
Comment (kailash-manasarovar) [inline docs/api/go/accounts.md:94]:
```suggestion
Complete examples are available on the [getting started](./getting-started.md) page.
```

File: docs/api/go/contract-interface.md [hunk 20]
```diff
@@ -0,0 +1,20 @@
+# Contract interfaces
+
+There is a set of system contract that helps with the interaction with zkSync Era network.<br>
```
Comment (kailash-manasarovar) [inline docs/api/go/contract-interface.md:3]:
```suggestion
There is a set of system contracts that supports the zkSync Era network interaction.<br>
```

File: docs/api/go/contract-interface.md [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+The following are the contracts:
+
+- ERC20
+- ContractDeployer
+
+### ERC20
+
+This is a system contract that provides management of ERC20 tokens. It has the same API as
```
Comment (kailash-manasarovar) [inline docs/api/go/contract-interface.md:11]:
```suggestion
This is a system contract that provides ERC20 token management. It has the same API as
```

File: docs/api/go/contract-interface.md [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+defined in [ERC20](https://eips.ethereum.org/EIPS/eip-20) standard.
```
Comment (kailash-manasarovar) [inline docs/api/go/contract-interface.md:12]:
```suggestion
defined in the [ERC20](https://eips.ethereum.org/EIPS/eip-20) standard.
```

File: docs/api/go/contract-interface.md [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+
+### ContractDeployer
+
+ContractDeployer is a utility system contract, represented as a type to cover the following functionality:
```
Comment (kailash-manasarovar) [inline docs/api/go/contract-interface.md:16]:
```suggestion
ContractDeployer is a utility system contract, represented as a type, and covers the following functionality:
```

File: docs/api/go/contract-interface.md [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+
+- encode binary contract representation by `create` method for further deploying.
```
Comment (kailash-manasarovar) [inline docs/api/go/contract-interface.md:18]:
```suggestion
- Encode binary contract representation by `create` method for further deploying.
```

File: docs/api/go/contract-interface.md [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+- encode binary contract representation by `create2` method for further deploying.
```
Comment (kailash-manasarovar) [inline docs/api/go/contract-interface.md:19]:
```suggestion
- Encode binary contract representation by `create2` method for further deploying.
```

File: docs/api/go/providers.md [hunk 26]
```diff
@@ -0,0 +1,62 @@
+# Providers
+
+Providers (clients) are objects that wrap interactions with the zkSync node. If you are new to the concept of providers,
```
Comment (kailash-manasarovar) [inline docs/api/go/providers.md:3]:
```suggestion
Providers (clients) are objects that wrap interactions with the zkSync Era node. If you are new to the concept of providers, 
```

File: docs/api/go/providers.md [hunk 27]
```diff
@@ -0,0 +1,62 @@
+# Providers
+
+Providers (clients) are objects that wrap interactions with the zkSync node. If you are new to the concept of providers, 
+you should check out their docs [here](https://geth.ethereum.org/docs/developers/dapp-developer/native).
```
Comment (kailash-manasarovar) [inline docs/api/go/providers.md:4]:
```suggestion
check out the [go-ethereum documentation](https://geth.ethereum.org/docs/developers/dapp-developer/native).
```

File: docs/api/go/providers.md [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+
+zkSync fully supports Ethereum Web3 API, so you can use the provider objects from `geth`. However, zkSync API provides some additional JSON-RPC methods, which allow:
+
+- Easily track L1<->L2 transactions.
```
Comment (kailash-manasarovar) [inline docs/api/go/providers.md:8]:
```suggestion
- Easy tracking of L1<->L2 transactions.
```

File: docs/api/go/providers.md [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+- Different stages of finality for transactions. By default, our RPC returns information about the last state processed by the server, but some use cases may require tracking "finalized" transactions only.
+
+`DefaultProvider` implements zkSync Web3 API and provides interaction with zkSync node.
```
Comment (kailash-manasarovar) [inline docs/api/go/providers.md:11]:
```suggestion
`DefaultProvider` implements the zkSync Era Web3 JSON-RPC API and provides interaction with zkSync node. 
```

File: docs/api/go/types.md [hunk 30]
```diff
@@ -0,0 +1,162 @@
+# Types
+
+All the types which are used in the SDK are referenced here
```
Comment (kailash-manasarovar) [inline docs/api/go/types.md:3]:
```suggestion
All the types which are used in the SDK are referenced
```

File: docs/api/go/types.md [hunk 31]
```diff
@@ -0,0 +1,162 @@
+# Types
+
+All the types which are used in the SDK are referenced here 
+(for more details check the [source code](https://github.com/zksync-sdk/zksync2-go/blob/main/Types.go)):
```
Comment (kailash-manasarovar) [inline docs/api/go/types.md:4]:
```suggestion
[in the source code](https://github.com/zksync-sdk/zksync2-go/blob/main/Types.go):
```

**Comment by uF4No** [comment]
@danijelTxFusion can you also add a "Requirements" section with any requirements like the minimal version of Go required, and a link to an installation guide?

**Comment by danijelTxFusion** [comment]
> @danijelTxFusion can you also add a "Requirements" section with any requirements like the minimal version of Go required, and a link to an installation guide?

It's already added [here](https://github.com/matter-labs/zksync-web-era-docs/pull/501/files#diff-b11d253325155215fb656e1c00101d772bb260546381a4d152c645205e94eb5aR16).

File: docs/api/go/accounts.md [hunk 32]
```diff
@@ -0,0 +1,95 @@
...
+func NewEthSignerFromMnemonic(mnemonic string, chainId int64) (*DefaultEthSigner, error)
+
+func NewEthSignerFromMnemonicAndAccountId(mnemonic string, accountId uint32, chainId int64) (*DefaultEthSigner, error)
+
+func NewEthSignerFromRawPrivateKey(rawPk []byte, chainId int64) (*DefaultEthSigner, error)
+```
+
+```go
+PrivateKey     := os.Getenv("PRIVATE_KEY")
+ZkSyncProvider := "https://zksync2-testnet.zksync.dev"
```
Comment (uF4No) [inline docs/api/go/accounts.md:20]:
```suggestion
ZkSyncProvider := "hhttps://testnet.era.zksync.dev"
```

File: docs/api/go/providers.md [hunk 33]
```diff
@@ -0,0 +1,62 @@
...
+
+## DefaultProvider
+
+Init:
+```go
+func NewDefaultProvider(rawUrl string) (*DefaultProvider, error)
+```
+
+```go
+ZkSyncProvider   := "https://zksync2-testnet.zksync.dev"
```
Comment (uF4No) [inline docs/api/go/providers.md:21]:
```suggestion
ZkSyncProvider   := "https://testnet.era.zksync.dev"
```

File: docs/api/go/getting-started.md [hunk 34]
```diff
@@ -0,0 +1,769 @@
...
+```console
+ go get github.com/zksync-sdk/zksync2-go
+```
+
+## Connecting to zkSync
+
+Once you have all the necessary dependencies, connect to zkSync using the endpoint of the operator node.
+
+```go
+ZkSyncProvider := "https://zksync2-testnet.zksync.dev"
```
Comment (uF4No) [inline docs/api/go/getting-started.md:31]:
```suggestion
ZkSyncProvider := "https://testnet.era.zksync.dev"
```

File: docs/api/go/getting-started.md [hunk 35]
```diff
@@ -0,0 +1,769 @@
...
+	"github.com/zksync-sdk/zksync2-go"
+	"log"
+	"math/big"
+	"os"
+)
+
+func main() {
+	var (
+		PrivateKey       = os.Getenv("PRIVATE_KEY")
+		ZkSyncProvider   = "https://zksync2-testnet.zksync.dev"
```
Comment (uF4No) [inline docs/api/go/getting-started.md:66]:
```suggestion
		ZkSyncProvider   = "https://testnet.era.zksync.dev"
```

File: docs/api/go/getting-started.md [hunk 36]
```diff
@@ -0,0 +1,769 @@
...
+	w, err := zksync2.NewWallet(es, zp)
+	if err != nil {
+		log.Panic(err)
+	}
+
+  // Get balance
+	balance, err := w.GetBalance()
+	if err != nil {
+		log.Panic(err)
+	}
```
Comment (uF4No) [inline docs/api/go/getting-started.md:195]:
```suggestion
  // Get balance
	balance, err := w.GetBalance()
	if err != nil {
		log.Panic(err)
	}
	fmt.Println("Balance is: ", balance)
```

File: docs/api/go/getting-started.md [hunk 37]
```diff
@@ -0,0 +1,770 @@
...
+
+There is a [user guide](https://github.com/zksync-sdk/zksync2-examples/blob/main/solidity/README.md) on how to compile Solidity smart contracts using `zksolc` 
+compiler. `zksolc` compiler generates a `*.zbin` and a `combined.json` file that contains the bytecode and ABI of a smart contract. 
+The `combined.json` file is used by `abigen` tool to generate smart contrat bindings. 
+Those files are used in the following examples. 
+
+
+#### Deploy a contract
+
+```
```
Comment (uF4No) [inline docs/api/go/getting-started.md:394]:
```suggestion
```go
```

**Comment by uF4No** [comment]
@danijelTxFusion as discussed, can you add the example for `ZksGetConfirmedTokens` in the getting started page?

File: docs/api/go/getting-started.md [hunk 38]
```diff
@@ -0,0 +1,803 @@
...
+defer zp.Close()
+```
+
+## Examples
+
+Full code for all examples is available [here](https://github.com/zksync-sdk/zksync2-examples/tree/main/go). Examples are configured to interact with `zkSync` and `Goerli` test networks.  
+
+### Get all confirmed tokens on zkSync
+
+Following example shows how to get all confirmed ERC20 tokens on zkSync network.
```
Comment (uF4No) [inline docs/api/go/getting-started.md:48]:
```suggestion
Following example shows how to get all confirmed tokens on zkSync Era. 
```

**Review by uF4No — APPROVED** [review]

**Review by uF4No — APPROVED** [review]
LGTM

**Comment by kailash-manasarovar** [comment]
Before you merge this PR, you need to create the menu links.

**Comment by danijelTxFusion** [comment]
> Before you merge this PR, you need to create the menu links.

I guess they are already added [here](https://github.com/matter-labs/zksync-web-era-docs/pull/501/files#diff-299095e13562e6413bead3ee43d199dcc66f92a4aa18b1442d78c921250d39d1R151).

**Review by kailash-manasarovar — APPROVED** [review]

**Pull request merged**